Add pigpio

joelbyler opened the pull request:
Looking to start using pigpio on the rpi0 and thought it would be good to try and get this into the main rpi0 system so others can take advantage.  If this looks good I'll work on creating an elixir library to build on top of this.  Info available at: http://abyz.co.uk/rpi/pigpio/

tmecklem committed:
Update CHANGELOG.md

fhunleth commented:
Let's get it here first, but I'd really like to upstream this to Buildroot. Here's a link to the previous comments to that pigpio patch for reference: http://patchwork.ozlabs.org/patch/717153/.

I'm quite fine with ignoring Python support. That will also make it easier to get into Buildroot.

fhunleth commented on package/pigpio/pigpio.mk:
```diff
@@ -0,0 +1,20 @@
+PIGPIO_LICENSE_FILES = UNLICENCE
+
+define PIGPIO_BUILD_CMDS
+	$(MAKE) $(TARGET_CONFIGURE_OPTS) VC_DIR=$(STAGING_DIR)/usr -C $(@D) -j4
```
Remove the `-j4` here.

fhunleth commented on package/pigpio/Config.in:
```diff
@@ -0,0 +1,6 @@
+config BR2_PACKAGE_PIGPIO
+	bool "pigpio"
```
Add this:
```
        depends on BR2_USE_MMU # fork()
        depends on BR2_TOOLCHAIN_HAS_THREADS
        depends on !BR2_STATIC_LIBS
 ```

fhunleth commented on package/pigpio/Config.in:
```diff
@@ -0,0 +1,6 @@
+	help
+	  PIGPIO daemon.
+
+	  https://github.com/joan2937/pigpio
```
Add this here:
```
comment "pigpio needs a toolchain w/ threads, dynamic library"                                                                          
        depends on !BR2_TOOLCHAIN_HAS_THREADS || BR2_STATIC_LIBS
```

fhunleth commented on package/pigpio/pigpio.mk:
```diff
@@ -0,0 +1,20 @@
+PIGPIO_SITE = $(call github,joan2937,pigpio,$(PIGPIO_VERSION))
+PIGPIO_LICENSE = Unlicense
+PIGPIO_LICENSE_FILES = UNLICENCE
+
```
Add this here:
```
PIGPIO_INSTALL_STAGING = YES 
```

fhunleth commented on package/pigpio/pigpio.mk:
```diff
@@ -0,0 +1,20 @@
+	$(MAKE) $(TARGET_CONFIGURE_OPTS) VC_DIR=$(STAGING_DIR)/usr -C $(@D) -j4
+endef
+
+define PIGPIO_INSTALL_TARGET_CMDS
```
Update the install commands to this:
```
define PIGPIO_INSTALL_STAGING_CMDS
        $(MAKE) $(TARGET_CONFIGURE_OPTS) prefix=/usr DESTDIR=$(STAGING_DIR) -C $(@D) install
endef                     
                          
define PIGPIO_INSTALL_TARGET_CMDS
        $(MAKE) $(TARGET_CONFIGURE_OPTS) prefix=/usr DESTDIR=$(TARGET_DIR) -C $(@D) install
endef                     
```
This will put everything in `/usr` instead of `/usr/local` which will be much more convenient. Buildroot will also automatically clean up the installation of unneeded files now. This will remove the man pages and headers files that currently are in the image.

fhunleth commented:
Could you also add a file `pigpio.hash` and put this into it:
```
# Locally calculated
sha256 3e0a416c993858123f21d873be97fde2e10bf8a6e9e53f69f0e12575f744a438  pigpio-4862a16c9f76f9b2096055c98ef4fbc480dc1878.tar.gz
```

joelbyler commented:
@fhunleth thanks for the feedback, I'll make the changes and test things out before moving forward.

fhunleth committed:
Force nerves_system_br 0.11.1 for erlinit fix

This is needed to fix an issue with shutting down and rebooting for
firmware updates.

fhunleth committed:
v0.13.1

fhunleth commented on package/pigpio/Config.in:
```diff
@@ -0,0 +1,13 @@
+  depends on BR2_USE_MMU # fork()
+  depends on BR2_TOOLCHAIN_HAS_THREADS
+  depends on !BR2_STATIC_LIBS
+
```
Remove the extra newline here. Sorry to be picky, but the Buildroot devs are super picky on these things.

fhunleth commented on package/pigpio/Config.in:
```diff
@@ -0,0 +1,13 @@
+  depends on !BR2_STATIC_LIBS
+
+	help
+	  PIGPIO daemon.
```
Could you change the help to "pigpio is a C library for the Raspberry Pi which allows control of the General Purpose Input Outputs (GPIO)."? Or if there's a better blurb on their website, use that.

joelbyler replied:
Yep, I'll look at the website and see if I can get a good description

fhunleth commented:
Added to nerves_system_br in https://github.com/nerves-project/nerves_system_br/commit/f2f75453c515fdc050534efd500022e19efb7b17

joelbyler committed:
enable pigpio for pi0

joelbyler committed:
picked up other changes from upstream

joelbyler committed:
missed update to mix.lock

fhunleth commented:
@tmecklem I suppose that you'd prefer future PRs be sent to https://github.com/nerves-project/nerves_system_rpi0, right? Anyway, have any issues with this one being included in the pi0? It's a 300KB feature that enables almost real-time gpio control.

@joelbyler Don't worry about resubmitting. Either Tim or I can merge it to the nerves-project/nerves_system_rpi0 repo fairly easily.

joelbyler commented:
@fhunleth need me to squash commits?

fhunleth commented:
Merged!  https://github.com/nerves-project/nerves_system_rpi0/commit/670bd346b61d3fca75421b0931aa0ceb9b771f82
